lib/ae_mdw/db/rocksdb.ex

```diff
@@ -0,0 +1,264 @@
+  - database
+  - the table references, one for each column family
+  - transaction
+  - iterator
+
+  Only the iterator is return by public function (the others are internal use only).
+  A column family iterator is created and released on demand (e.g. for next and prev after multiple iterations).
+
+  A transaction is automatically started by the first put and finishes with a commit.
+  On the next put a new transaction begins. In order to write directly, a dirty_put can be used.
+  """
+  alias AeMdw.Log
+  alias AeMdw.Db.Model
+  alias AeMdw.Db.Mutation
+
+  @opaque iterator :: reference() | binary()
+  # when transactions are explicit use:
+  # @opaque transaction :: reference() | binary()
+  @typep table :: atom()
+
+  # arround 50 column families * 10
+  @max_open_files 500
+  # default from https://github.com/facebook/rocksdb/wiki/Space-Tuning
+  @block_size 4 * 1024
+  @block_cache_size 32 * 1024 * 1024
+  @write_buffer_size 32 * 1024 * 1024
+
+  @db_options [
+    create_if_missing: true,
+    create_missing_column_families: true,
+    max_background_compactions: 4,
+    max_background_flushes: 2,
+    use_fsync: true,
+    bytes_per_sync: 1_048_576,
+    max_open_files: @max_open_files,
+    merge_operator: :erlang_merge_operator
+  ]
+  @cf_options [
+    # Mdw db is only 4% of total used storage
+    # (with cpu-optimized machines we can get probably faster writes by enabling it though)
+    compression: :none,
+    bottommost_compression: :none,
+    write_buffer_size: @write_buffer_size,
+    block_based_table_options: [
+      block_size: @block_size,
+      block_cache_size: @block_cache_size,
+      cache_index_and_filter_blocks: true,
+      format_version: 5
+    ],
+    level_compaction_dynamic_level_bytes: true,
+    merge_operator: :erlang_merge_operator
+  ]
+
+  @doc """
+  Opens an optimistic transaction database with multiple column families in order to allow transactions
+  across multiple tables.
+  """
+  @spec open() :: :ok
+  def open() do
+    with :ok <- create_dir_if_missing(),
+         {:ok, db_ref, cf_ref_list} <- open_db() do
+      # keep db handle
+      :persistent_term.put({__MODULE__, :db_ref}, db_ref)
+
+      # map each cf handle to its cf model
+      all_cf_names()
+      |> Enum.zip(cf_ref_list)
+      |> Enum.each(fn {cf_name, cf_ref} ->
+        :persistent_term.put({__MODULE__, cf_name}, {db_ref, cf_ref})
+      end)
+    else
+      {:error, reason} ->
+        Log.error("Failed to open database! path=#{data_dir()}, reason=#{inspect(reason)}")
+        :error
+    end
+  end
+
+  @doc """
+  Puts changes into an optimistic transaction.
+
+  RocksDB raises an exception on conflict which is helpful to identify unwanted parallelism in the future when
+  it doesn't respect dependency.
+  """
+  @spec put(table(), binary(), binary()) :: :ok | :error
+  def put(table, key, value) do
+    with {_db_ref, cf_ref} <- cf_refs(table),
+         {:ok, t_ref} <- get_transaction() do
```

Comment:
`get_transaction` will always return a transaction, no need for `{:ok, transaction}` or to include it inside the `with` statement